azure-kusto-data/src/connection_string.rs

```diff
@@ -135,267 +149,852 @@ static ALIAS_MAP: Lazy<HashMap<&'static str, ConnectionStringKey>> = Lazy::new(|
+                client_id,
+                client_authority,
+                thumbprint,
+                private_certificate_path,
+            } => {
+                write!(
+                    f,
+                    "ApplicationCertificate({}, {}, {}, {})",
+                    client_id,
+                    client_authority,
+                    thumbprint,
+                    private_certificate_path.display()
+                )
+            }
+            ConnectionStringAuth::ManagedIdentity { user_id } => {
+                write!(
+                    f,
+                    "ManagedIdentity({})",
+                    user_id.as_deref().unwrap_or("<none>")
+                )
+            }
+            ConnectionStringAuth::AzureCli => write!(f, "AzureCli"),
+            ConnectionStringAuth::DeviceCode { .. } => {
+                write!(f, "DeviceCode()")
+            }
+            ConnectionStringAuth::InteractiveLogin => write!(f, "InteractiveLogin"),
+            ConnectionStringAuth::TokenCredential { .. } => write!(f, "TokenCredential"),
+        }
+    }
+}
 
+impl ConnectionString {
+    /// Parses a connection string in order to create a `ConnectionString` object.
+    /// The connection string is a series of key-value pairs separated by semicolons.
+    /// # Examples
+    /// ```rust
+    /// use azure_kusto_data::error::Error;
+    ///
+    /// use azure_kusto_data::prelude::*;
+    /// # fn main() -> Result<(), Error> {
+    /// let connection_string = ConnectionString::from_raw_connection_string("Data Source=localhost ; Application Client Id=f6f295b1-0ce0-41f1-bba3-735accac0c69; Appkey =1234;Authority Id= 25184ef2-1dc0-4b05-84ae-f505bf7964f4 ; aad federated security = True")?;
+    ///
+    /// assert_eq!(connection_string.auth, ConnectionStringAuth::Application {
+    ///    client_id: "f6f295b1-0ce0-41f1-bba3-735accac0c69".to_string(),
+    ///   client_authority: "25184ef2-1dc0-4b05-84ae-f505bf7964f4".to_string(),
+    ///  client_secret: "1234".to_string(),
+    /// });
+    /// assert_eq!(connection_string.data_source, "localhost");
+    /// assert_eq!(connection_string.federated_security, true);
+    /// # Ok::<(), Error>(()) }
+    /// ```
+    pub fn from_raw_connection_string(
+        connection_string: &str,
+    ) -> Result<Self, ConnectionStringError> {
         let kv_str_pairs = connection_string
             .split(';')
             .filter(|s| !s.chars().all(char::is_whitespace));
 
+        let mut result_map = HashMap::<ConnectionStringKey, &str>::new();
+
         for kv_pair_str in kv_str_pairs {
             let mut kv = kv_pair_str.trim().split('=');
             let k = match kv.next().filter(|k| !k.chars().all(char::is_whitespace)) {
                 None => {
                     return Err(ConnectionStringError::ParsingError {
-                        msg: "No key found".to_owned(),
+                        msg: "No key found".to_string(),
```

Comment:
one of the instances where he casting in rust gets a bit annoying 😆 . Maybe in a follow up PR we want to indocude some convenience functions I have seem in other crate like `ConnectionStringError::parsing_error(..)`.

Reply:
Did it, is that what you had in mind?